### PR TITLE
[DML EP] Support partial rotary embedding

### PR DESCRIPTION
**Pull request by PatriceVignola**
### Description
This adds support for partial RotaryEmbedding to DML. Essentially, partial RotaryEmbedding simply consists of doing the rotary embedding calculation on a subregion of the input tensor of as if its head size was `rotary_embedding_dim`, while leaving the second part of the tensor (i.e. `head_size - rotary_embedding_dim`) alone.

To achieve this, all we need to do is follow the following steps:

1. Split the tensor into 2 parts
2. Run the rotary embedding algorithm on the first part, just like we were doing before on the entire tensor
3. Join the 2 parts back together

Since we're leaving the middle part intact, the RotaryEmbedding fusion will still be done within DML. Also, the concat at the end is essentially free because DML optimizes it out and directly allocate the result of RotaryEmbedding at the right place. The only overhead here is the splitting of the tensor at the beginning, which we should eventually make part of the RotaryEmbedding fusion within DML.

### Motivation and Context
This fix allows us to correctly run models that have a `partial_rotary_factor` setting in huggingface, including Nvidia's Nemotron: https://huggingface.co/nvidia/Nemotron-Mini-4B-Instruct

### CODE REVIEW

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp [hunk 1]
```diff
@@ -56,25 +56,45 @@
...
-        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[inputDataIndex].GetSizes().back() % headSize == 0);
+        uint32_t numHeads = gsl::narrow_cast<uint32_t>(kernelInfo.GetOptionalAttribute<int64_t>(AttrName::NumHeads, 0));
+        uint32_t rotaryEmbeddingDim = gsl::narrow_cast<uint32_t>(kernelInfo.GetOptionalAttribute<int64_t>(AttrName::RotaryEmbeddingDim, 0));
 
-        // We resize the data to be of shape [batchSize, sequenceLength, numHeads, headSize]
         const auto inputDataSizes = m_inputTensorDescs[inputDataIndex].GetSizes();
+        const uint32_t hiddenSize = inputIs4D ? inputDataSizes[1] * inputDataSizes[3] : inputDataSizes.back();
+
+        const uint32_t headSize = numHeads == 0
+            ? m_inputTensorDescs[cosCacheIndex].GetSizes().back() * 2
```
Comment (sumitsays) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp:69]: >cosCacheIndex [](http://example.com/codeflow?start=33&length=13)

Last dimension of `cos_cache` can be of `rotary_embedding_dim/2` as per the contrib op doc. So I think we should use last dimension of `input` tensor. (Based on assumption that when `num_heads` is 0, then `input` tensor will always be 4-D).

**Comment by sumitsays** [comment]
> 

If possible can we have the ASCII DML graph. It will be helpful to understand the implementation.

---
Refers to: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp:27 in 7370243. [](commit_id = 73702430ed23b70683b45c7177b02bb6e2e79db0, deletion_comment = False)

**Review by sumitsays — APPROVED** [review]
:shipit:

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp (hunk 1 above)
Comment (PatriceVignola) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp:69]: That's not how num_heads work (it's unrelated to whether the input is 3D or 4D). On the other hand, the last dimension of cos_cache will always be `headSize / 2` when `numHeads == 0`. Essentially, when `numHeads == 0`, it means that `rotaryEmbeddingDim == 0` and therefore we are in a situation where we have to use the entire cache. If `numHeads != 0`, then we have to use `hiddenSize / numHeads` to figure out the size of the rotary embedding dim.

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp [hunk 2]
```diff
@@ -25,6 +25,41 @@
...
+          |         |              |
+          |         +-----+ +------+
+          |               | |
+          |               Add
+          |                |
+          +-------------+  |
+                        |  |
+                        Join
+*/
+
```
Comment (sumitsays) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp:62]: Thank you for adding this!

**Review by sumitsays — APPROVED** [review]

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp (hunk 2 above)
Comment (fdwr) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp:62]: Yes, it's lovely and will be very helpful for those poor folks who have to understand it later (like me :)).

**Comment by snnn** [comment]
This PR has been cherry-picked into the `rel-1.20.0` branch in PR #22526. Removing the `release:1.20.0` label.

**Pull request merged**